README.md

```diff
@@ -52,13 +52,13 @@ packages:
   '@*/*':
     # scoped packages
     access: $all
-    publish: $authenticated
+    publish: $owned-group
```

Comment:
This should e.g. be `$maintainer `as described here https://github.com/bufferoverflow/verdaccio-gitlab#configuration-options.

Reply:
I thought about doing it this way but the issue is that I would need to retrieve all user's projets/groups. The current way is to retrieve only the groups/projets where the user is at least `auth.gitlab.publish` meaning that I would have to change that as well.

Reply:
Originally, when working on this PR, I decided to use `$owned-group` as a way to represent the current behavior that the plugin is using for `publish` authorization.
The naming may not be ideal (though I have nothing better in mind, yet) but it's there for the user to have a way to say "Only do this action if you have at least `auth.gitlab.publish` permission on the associated group/project".

I think adding support for `$maintainer`, `$owner`, ... beyond the `auth.gitlab.publish` should be another PR with a discussion regarding the design itself.

I'm not sure you remember my other issue but I made this PR because I plan to do #99 right after it's merged.

Reply:
@bufferoverflow What should I do regarding this comment?

Reply:
`$maintainer` is the current default as figured out [here](https://github.com/bufferoverflow/verdaccio-gitlab/pull/101/files#r355184353), so we do not need `$owned-group`

Reply:
Those two points are independent, I specifically said `If I were to do it the way you suggest` at the beginning of my message because it is not the way I designed it to work in this PR.

Sorry if I'm not clear enough, I'll try to better explain:

The main aim of this PR is to provide a slightly more intuitive/explicit configuration.  
The current behavior of the plugin is to do something very specific for `publish` and `access` authorizations. There is no way to customize the behavior: I can't allow `$anonymous` to `publish` and I can't enforce the current `publish` behavior for `access` authorization.  

This PR attempts to fix this issue with two reworks at once:
- Unify `allow_publish` and `allow_access` under a `allow_action` with specific arguments and defaults.
- Explicitly handle the different use-cases the user may be willing to use (for example `$anonymous` which allows anyone to perform the action or `$authenticated` that allow any authenticated user to perform the action)

Handling the first point was fairly easy but the second point requires more explicit configuration from the user. He can't simply use anything as the value of `publish` or `access`. He needs to state the exact behavior the module will use.

Finally, this stricter/more explicit handling of permissions created the need for some group to represent the old `publish` behavior. That's what `$owned-group` solves even though the name is not ideal.


I hope this explanation is clear.

Reply:
@bufferoverflow Any feedback regarding my explanation?

Reply:
I currently prefer separate handling instead of a unified function and I think caching should be enabled for this, otherwise you DDoS the gitlab server with each npm install.

Reply:
> I currently prefer separate handling instead of a unified function

Do you have any reason for this?

> I think caching should be enabled for this, otherwise you DDoS the gitlab server with each npm install.

Where do I call the gitlab server? I copy/pasted the previous code and unified it

Reply:
caching is done based on the setting within publish level, this is used within your code as well, so it only works if access and publish level is the same when cache is enabled.

Reply:
Do you mean that the `authenticate` function is called often?  
Even if that's the case, the user is cached with `_setCachedUserGroups` so it won't call the Gitlab API more than once per user.  
`access` and `publish` permissions both rely on `publishLevel` when `"$owned-group"` is specified.

Reply:
https://github.com/bufferoverflow/verdaccio-gitlab/blob/master/src/authcache.ts#L50 does not have the access level, so if we add functionality for different access levels for access, people should be able to define the level of their choice for access and publish.

Reply:
Yes, for now they will both use the `publishLevel` value without distinction.  
The proper way would be to be able to specify `$maintainer`, `$owner` and so on for both `publish` and `access` but this requires much more work. (as mentioned here: https://github.com/bufferoverflow/verdaccio-gitlab/pull/101#discussion_r355183736)

Reply:
Do you have any other feedback or is this PR ok as-is?

Reply:
as said, **if we add functionality for different access levels, people should be able to define the level of their choice for access and publish.**

Reply:
Yes but this is not the goal of this PR. Its only goal is to increase security.